### PR TITLE
refactor(youtube/general-ads-patch): hide endscreen patch

### PR DESCRIPTION
**Pull request by ghost**
https://github.com/revanced/revanced-integrations/pull/200

### CODE REVIEW

File: src/main/kotlin/app/revanced/patches/youtube/ad/general/bytecode/patch/GeneralBytecodeAdsPatch.kt [hunk 1]
```diff
@@ -248,22 +249,27 @@ class GeneralBytecodeAdsPatch : BytecodePatch() {
...
-                                    if (invokeInstruction.opcode != Opcode.IPUT_OBJECT) return@forEachIndexed
+                                    //  and is followed by an instruction with the mnemonic CHECK_CAST
+                                    var insertIndex = index + 4
+                                    var checkCastInstruction = instructions.elementAt(insertIndex)
+                                    if (checkCastInstruction.opcode != Opcode.CHECK_CAST) {
+                                        // perform a new check for endscreen_element_layout_icon
+                                        insertIndex -= 1
+                                        checkCastInstruction = instructions.elementAt(insertIndex)
+
+                                        if (checkCastInstruction.opcode != Opcode.CHECK_CAST) return@forEachIndexed
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/youtube/ad/general/bytecode/patch/GeneralBytecodeAdsPatch.kt:260]: Here should be returned a PatchResult.Error instance.

File: src/main/kotlin/app/revanced/patches/youtube/ad/general/bytecode/patch/GeneralBytecodeAdsPatch.kt [hunk 2]
```diff
@@ -248,22 +249,27 @@ class GeneralBytecodeAdsPatch : BytecodePatch() {
                                 }
 
                                 resourceIds[3], resourceIds[4], resourceIds[5] -> { // end screen ads
-                                    //  and is followed by an instruction with the mnemonic IPUT_OBJECT
-                                    val insertIndex = index + 7
-                                    val invokeInstruction = instructions.elementAt(insertIndex)
-                                    if (invokeInstruction.opcode != Opcode.IPUT_OBJECT) return@forEachIndexed
+                                    //  and is followed by an instruction with the mnemonic CHECK_CAST
+                                    var insertIndex = index + 4
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/youtube/ad/general/bytecode/patch/GeneralBytecodeAdsPatch.kt:253]: Use a subList view of the current instruction list to find the insertion index instead of checking two offsets like this.

**Review by oSumAtrIX — CHANGES_REQUESTED** [review]

**Review by oSumAtrIX — CHANGES_REQUESTED** [review]
Seperate hiding endscreens into a seperate parch. It is not an ad.

**Comment by OxrxL** [comment]
> Seperate hiding endscreens into a seperate parch. It is not an ad.

Done

File: src/main/kotlin/app/revanced/patches/youtube/layout/hideendscreencards/bytecode/patch/HideEndscreenCardsPatch.kt [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+@DependsOn([IntegrationsPatch::class, HideEndscreenCardsResourcePatch::class])
+@Name("hide-endscreen-cards-patch")
+@Description("Hides the suggested video cards at the end of a video in fullscreen.")
+@HideEndScreenCardsCompatibility
+@Version("0.0.1")
+class HideEndscreenCardsPatch : BytecodePatch(
+    listOf(
+        LayoutCircleFingerprint,
+        LayoutIconFingerprint,
+        LayoutVideoFingerprint,
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/youtube/layout/hideendscreencards/bytecode/patch/HideEndscreenCardsPatch.kt:33]: All of these fingerprints are scanned for through all classes. Instead, create 1 fingerprint which you use to find the method, then create three more fingerprints which you can resolve on to the first fingerprint. This way you can get the offsets without having to find the method in all classes.

File: src/main/kotlin/app/revanced/patches/youtube/layout/hideendscreencards/bytecode/patch/HideEndscreenCardsPatch.kt [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    )
+) {
+    override fun execute(context: BytecodeContext): PatchResult {
+        fun injectInvokeCall(result: MethodFingerprintResult) {
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/youtube/layout/hideendscreencards/bytecode/patch/HideEndscreenCardsPatch.kt:37]: Use this signature to reduce duplicate code:
```suggestion
        fun MethodFingerprint.injectHideCall() {
```

File: src/main/kotlin/app/revanced/patches/youtube/layout/hideendscreencards/bytecode/patch/HideEndscreenCardsPatch.kt [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+            val layoutMethod = result.mutableMethod
+
+            val checkCastIndex = result.scanResult.patternScanResult!!.endIndex
+            val viewRegister = (layoutMethod.instruction(checkCastIndex) as Instruction21c).registerA
+
+            layoutMethod.addInstruction(
+                checkCastIndex + 1,
+                "invoke-static { v$viewRegister }, Lapp/revanced/integrations/patches/HideEndscreenPatch;->HideEndscreen(Landroid/view/View;)V"
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/youtube/layout/hideendscreencards/bytecode/patch/HideEndscreenCardsPatch.kt:45]: This method reference is out of sync with the integrations, please test your changes before marking a PR as ready.

**Review by oSumAtrIX — CHANGES_REQUESTED** [review]

File: src/main/kotlin/app/revanced/patches/youtube/layout/hideendscreencards/bytecode/patch/HideEndscreenCardsPatch.kt (hunk 3 above)
Comment (OxrxL) [reply src/main/kotlin/app/revanced/patches/youtube/layout/hideendscreencards/bytecode/patch/HideEndscreenCardsPatch.kt:33]: I don't think it is useful to do this. They are 3 methods of 3 different classes, they are not all in one class.

File: src/main/kotlin/app/revanced/patches/youtube/layout/hideendscreencards/resource/patch/HideEndscreenCardsResourcePatch.kt [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+        layoutCircle = ResourceMappingResourcePatch.resourceMappings.single {
+            it.type == "layout" && it.name == "endscreen_element_layout_circle"
+        }.id
+        layoutIcon = ResourceMappingResourcePatch.resourceMappings.single {
+            it.type == "layout" && it.name == "endscreen_element_layout_icon"
+        }.id
+        layoutVideo = ResourceMappingResourcePatch.resourceMappings.single {
+            it.type == "layout" && it.name == "endscreen_element_layout_video"
+        }.id
```
[the remark this answers is not in the document]
Comment (OxrxL) [reply src/main/kotlin/app/revanced/patches/youtube/layout/hideendscreencards/resource/patch/HideEndscreenCardsResourcePatch.kt:46]: There's no example of this between patches.

[the remark this answers is not in the document]
Comment (oSumAtrIX) [reply src/main/kotlin/app/revanced/patches/youtube/layout/hideendscreencards/resource/patch/HideEndscreenCardsResourcePatch.kt:46]:
```kotlin
fun find(name: String) = ResourceMappingResourcePatch.resourceMappings.single {
       it.type == "layout" && it.name == name
}.id
```

File: src/main/kotlin/app/revanced/patches/youtube/layout/hideendscreencards/bytecode/patch/HideEndscreenCardsPatch.kt [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+            layoutMethod.addInstruction(
+                checkCastIndex + 1,
+                "invoke-static { v$viewRegister }, Lapp/revanced/integrations/patches/HideEndscreenCardsPatch;->hideEndscreen(Landroid/view/View;)V"
+            )
+        }
+
+        injectInvokeCall(LayoutCircleFingerprint)
+        injectInvokeCall(LayoutIconFingerprint)
+        injectInvokeCall(LayoutVideoFingerprint)
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/youtube/layout/hideendscreencards/bytecode/patch/HideEndscreenCardsPatch.kt:55]:
```suggestion
        fun MethodFingerprint.injectHideCall() {
            val layoutResult = result!!
            val layoutMethod = layoutResult.mutableMethod

            val checkCastIndex = layoutResult.scanResult.patternScanResult!!.endIndex
            val viewRegister = (layoutMethod.instruction(checkCastIndex) as Instruction21c).registerA

            layoutMethod.addInstruction(
                checkCastIndex + 1,
                "invoke-static { v$viewRegister }, Lapp/revanced/integrations/patches/HideEndscreenCardsPatch;->hideEndscreen(Landroid/view/View;)V"
            )
        }
        
        listOf(LayoutCircleFingerprint, LayoutIconFingerprint, LayoutVideoFingerprint).forEach(::injectHideCall())
```

**Review by oSumAtrIX — CHANGES_REQUESTED** [review]
Looks good, why is it necessary to patch three places to hide the end cards. Instead can you not hide the view, to which they are added to?

**Comment by OxrxL** [comment]
> Looks good, why is it necessary to patch three places to hide the end cards. Instead can you not hide the view, to which they are added to?

Because this was the patch originally developed by Vanced. It works and I have therefore never changed it (based on the ancient rule: if it's not broken, don't fix it).

**Comment by oSumAtrIX** [comment]
Build seems to fail: https://github.com/revanced/revanced-patches/actions/runs/3414296627/jobs/5682064547.

Please resolve the issue and we can get it merged!

**Comment by OxrxL** [comment]
> Build seems to fail: https://github.com/revanced/revanced-patches/actions/runs/3414296627/jobs/5682064547.
> 
> Please resolve the issue and we can get it merged!

It's caused by your suggestions. I don't know how to fix it.

**Comment by OxrxL** [comment]
Immagine 2022-11-07 224731

**Comment by oSumAtrIX** [comment]
These have been forgotten to be removed:

https://github.com/revanced/revanced-patches/blob/main/src/main/kotlin/app/revanced/patches/youtube/ad/general/bytecode/patch/GeneralBytecodeAdsPatch.kt#L246

File: src/main/kotlin/app/revanced/patches/youtube/ad/general/bytecode/patch/GeneralBytecodeAdsPatch.kt [hunk 8]
```diff
@@ -227,41 +223,6 @@ class GeneralBytecodeAdsPatch : BytecodePatch() {
...
-                                    if (invokeInstruction.opcode != Opcode.INVOKE_VIRTUAL) return@forEachIndexed
-
-                                    // create proxied method, make sure to not re-resolve() the current class
-                                    if (mutableClass == null) mutableClass = context.proxy(classDef).mutableClass
-                                    if (mutableMethod == null) mutableMethod =
-                                        mutableClass!!.findMutableMethodOf(method)
-
-                                    //ToDo: Add Settings toggle for whatever this is
-                                    mutableMethod!!.implementation!!.removeInstruction(removeIndex)
-                                }
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/youtube/ad/general/bytecode/patch/GeneralBytecodeAdsPatch.kt:244]: Please open a separate PR for this code block and the corresponding resource id. It is not part of this PR.

**Review by oSumAtrIX — CHANGES_REQUESTED** [review]

**Review by oSumAtrIX — CHANGES_REQUESTED** [review]
